aptos-move/block-executor/src/executor.rs

```diff
@@ -1129,9 +1105,18 @@ where
             AptosModuleExtension,
         >,
         runtime_environment: &RuntimeEnvironment,
-        total_txns_to_materialize: &AtomicU32,
         final_results: &ExplicitSyncWrapper<Vec<E::Output>>,
+        output_idx: Option<TxnIndex>,
     ) -> Result<(), PanicError> {
+        if output_idx.is_some_and(|idx| idx < txn_idx) {
+            return Err(code_invariant_error(format!(
+                "Index to record finalized output {} is less than txn index {}",
+                output_idx.unwrap(),
+                txn_idx
+            )));
+        }
+        let output_idx = output_idx.unwrap_or(txn_idx);
```

Comment:
@grao1991 : output_idx - I should rename to maybe_output_idx, is provided by materialize_txn_commit call from finalize_parallel_execution (above we check invariant it can't be lower than the index at which it executed) and then set in final outputs at that (output) index. If output idx isn't provided, then we take txn_idx, as is for materialize in the normal worker loop flow.